### PR TITLE
tango-controls: init at 9.4.2

### PR DESCRIPTION
**Pull request by pmiddend**
## Description of changes

This adds the [Tango](https://www.tango-controls.org/) controls system that is being widely used by different particle accelerator facilities across the globe. We're using it here at [DESY](https://www.desy.de/) to control our motors, detectors etc., and we're using Nix as well, so it makes sense to put it into nixpkgs.

This derivation is sort of battle-tested by now, but it's hard to give an example. You could build the derivation and try to run one of the Java-based UIs for Tango, Jive: `result/bin/jive`. This needs a working Tango server, however, and that's not trivial to set up because it needs a MariaDB database.

I have systemd services for that as well, but I wanted to contribute just the core.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pmiddend** [comment]
@ofbord build tango-controls

**Review by Samy32377777 — APPROVED** [review]

File: pkgs/applications/system/tango-controls/default.nix [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+, pkg-config
+, python3
+, stdenv
+, systemd
+, zeromq
+, zlib
+}:
+let
+  # omniorb is in nixpkgs, but tango doesn't support the latest version
+  omniorb_4_2 = stdenv.mkDerivation rec {
```
Comment (natsukium) [inline pkgs/applications/system/tango-controls/default.nix:20]: How about overriding?

File: pkgs/applications/system/tango-controls/default.nix [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    "-DMySQL_EXECUTABLE=${mariadb-connector-c}/bin/mariadb"
+    "-DCMAKE_VERBOSE_MAKEFILE=TRUE"
+    "-DCMAKE_SKIP_RPATH=ON"
+    "-DTSD_TAC=ON"
+    "-DTSD_JAVA_PATH=${openjdk11}"
+  ];
+
+  # To build proper systemd services for Tango, we need a way to tell when the DB service is started.
+  # This patch adds systemd support for this.
+  patches = lib.optional (!stdenv.isDarwin) ./sd_notify_cmake.patch;
```
Comment (natsukium) [inline pkgs/applications/system/tango-controls/default.nix:73]:
```suggestion
  patches = lib.optional stdenv.isLinux ./sd_notify_cmake.patch;
```
For example, BSD doesn't support systemd.

File: pkgs/applications/system/tango-controls/default.nix [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    done
+  '';
+
+  meta = with lib; {
+    description = "Open Source solution for SCADA and DCS";
+    longDescription = ''
+      Tango Controls is a toolkit for connecting hardware and software together. It is a mature software which is used by tens of sites to run highly complicated accelerator complexes and experiments 24 hours a day. It is free and Open Source. It is ideal for small and large installations. It provides full support for 3 programming languages - C++, Python and Java.
+    '';
+    homepage = "https://www.tango-controls.org";
+    changelog = "https://gitlab.com/tango-controls/cppTango/-/blob/main/RELEASE_NOTES.md";
```
Comment (natsukium) [inline pkgs/applications/system/tango-controls/default.nix:94]:
```suggestion
    changelog = "https://gitlab.com/tango-controls/cppTango/-/blob/${version}/RELEASE_NOTES.md";
```

File: pkgs/applications/system/tango-controls/default.nix [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    enableParallelBuilding = true;
+    hardeningDisable = [ "format" ];
+  };
+in
+stdenv.mkDerivation rec {
+  pname = "tango-controls";
+  version = "9.4.2";
+
+  src = fetchurl {
```
Comment (natsukium) [inline pkgs/applications/system/tango-controls/default.nix:40]: Why don't you use fetchFromGitLab?

Comment (pmiddend) [reply pkgs/applications/system/tango-controls/default.nix:34]: Very good question! The thing about tango-controls is that it's assembled from a lot of components, all in different repositories: [starter](https://gitlab.com/tango-controls/starter), [cpptango](https://gitlab.com/tango-controls/cppTango), [tango-database](https://gitlab.com/tango-controls/TangoDatabase).

The `TangoSourceDistribution` repository contains an `ant` build script to clone all these repositories and build them. It also contains a package (the tarball that I download here) that has all these repositories with the proper versions (!) pre-cloned, so building is _much_ easier.

File: pkgs/applications/system/tango-controls/default.nix (hunk 3 above)
Comment (pmiddend) [reply pkgs/applications/system/tango-controls/default.nix]: Fixed, thanks!

File: pkgs/applications/system/tango-controls/default.nix (hunk 2 above)
Comment (pmiddend) [reply pkgs/applications/system/tango-controls/default.nix]: Good point, fixed!

File: pkgs/applications/system/tango-controls/default.nix (hunk 1 above)
Comment (pmiddend) [reply pkgs/applications/system/tango-controls/default.nix]: Yep, shortens the derivation a bit, fixed!

**Comment by pmiddend** [comment]
I'll squash once the review is over.

**Comment by pmiddend** [comment]
I have to give this up, since my day job doesn't involve tango anymore.

**Pull request closed without merging**